src/blocks/scratch3_event.js

```diff
@@ -12,23 +12,76 @@ function Scratch3EventBlocks(runtime) {
  */
 Scratch3EventBlocks.prototype.getPrimitives = function() {
     return {
-        'event_whenflagclicked': this.whenFlagClicked,
-        'event_whenbroadcastreceived': this.whenBroadcastReceived,
-        'event_broadcast': this.broadcast
+        'event_broadcast': this.broadcast,
+        'event_broadcastandwait': this.broadcastAndWait,
+        'event_whengreaterthan': this.hatGreaterThanPredicate
     };
 };
 
+Scratch3EventBlocks.prototype.getHats = function () {
+    return {
+        'event_whenflagclicked': {
+            restartExistingThreads: true
+        },
+        /*'event_whenkeypressed': {
+            restartExistingThreads: false
+        },
+        'event_whenthisspriteclicked': {
+            restartExistingThreads: true
+        },
+        'event_whenbackdropswitchesto': {
+            restartExistingThreads: true
+        },*/
+        'event_whengreaterthan': {
+            restartExistingThreads: false,
+            edgeTriggered: true
+        },
+        'event_whenbroadcastreceived': {
+            restartExistingThreads: true
+        }
+    };
+};
 
-Scratch3EventBlocks.prototype.whenFlagClicked = function() {
-    // No-op
+Scratch3EventBlocks.prototype.hatGreaterThanPredicate = function (args, util) {
+    // @todo: Other cases :)
+    if (args.WHENGREATERTHANMENU == 'TIMER') {
+        return util.ioQuery('clock', 'projectTimer') > args.VALUE;
+    }
+    return false;
 };
 
-Scratch3EventBlocks.prototype.whenBroadcastReceived = function() {
-    // No-op
+Scratch3EventBlocks.prototype.broadcast = function(args, util) {
+    util.triggerHats('event_whenbroadcastreceived', {
+        'BROADCAST_OPTION': args.BROADCAST_OPTION
+    });
 };
 
-Scratch3EventBlocks.prototype.broadcast = function() {
-    // @todo
+Scratch3EventBlocks.prototype.broadcastAndWait = function (args, util) {
+    // Have we run before, triggering threads?
+    if (!util.stackFrame.triggeredThreads) {
+        // No - trigger hats for this broadcast.
+        util.stackFrame.triggeredThreads = util.triggerHats(
+            'event_whenbroadcastreceived', {
+                'BROADCAST_OPTION': args.BROADCAST_OPTION
+            }
+        );
+        if (util.stackFrame.triggeredThreads.length == 0) {
+            // Nothing was started.
+            return;
+        }
+    }
+    // We've run before; check if the wait is still going on.
+    var waiting = false;
+    for (var i = 0; i < util.stackFrame.triggeredThreads.length; i++) {
+        var thread = util.stackFrame.triggeredThreads[i];
+        var activeThreads = this.runtime.threads;
+        if (activeThreads.indexOf(thread) > -1) { // @todo: A cleaner way?
+            waiting = true;
+        }
+    }
```

Comment:
yep, that would be cleaner! you are right though that those are ES6 features. maybe once we do #3.

I would like it though if the block implementations didn't need to access `runtime`. We are thinking of the block implementations as eventually something that could be stuck into Scratch extensions, so trying to avoid giving them too much privileged interpreter info. But, this seems like an especially tricky case.

We are using promises to let particular blocks "pause" their thread's execution while waiting for certain results. E.g., our color-touching-color implementation might require crossing the WebWorker thread boundary, so it can return a promise and give its result when the answer comes back across the thread.